### PR TITLE
refactor: remove parallel walk

### PR DESCRIPTION
**Pull request by knqyf263**
## Description
Since filepath.WalkDir is fast enough now, this PR removes `github.com/saracen/walker`. Also, it defines the `Walker` interface so that other tools can override the behavior. 

## Related issues
- Close https://github.com/aquasecurity/trivy/issues/5179

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

**Comment by knqyf263** [comment]
@santhosh1729 I think we'll merge this PR now, and you can give us feedback anytime.

**Review by tonaim — APPROVED** [review]

**Comment by knqyf263** [comment]
@DmitriyLewen Could you quickly take a look? Aqua confirmed it, and it should be fine, but I want to double-check it.

File: pkg/fanal/walker/fs_test.go [hunk 1]
```diff
@@ -2,8 +2,11 @@ package walker_test
 
 import (
 	"errors"
+	"golang.org/x/exp/slices"
```
Comment (nikpivkin) [inline pkg/fanal/walker/fs_test.go:5]:
```suggestion
	"slices"
```

By the way, we use "golang.org/x/exp/maps" and "golang.org/x/exp/slices" in many places, but we can already switch to the standard slices and maps packages.

File: pkg/fanal/walker/vm.go [hunk 2]
```diff
@@ -34,22 +34,22 @@ func AppendPermitDiskName(s ...string) {
...
-		threshold: threshold,
+		logger: log.WithPrefix("vm"),
 	}
 }
 
-func (w *VM) Walk(vreader *io.SectionReader, root string, fn WalkFunc) error {
+func (w *VM) Walk(vreader *io.SectionReader, root string, opt Option, fn WalkFunc) error {
+	w.skipFiles = opt.SkipFiles
+	w.skipDirs = append(w.skipDirs, defaultSkipDirs...)
+
```
Comment (nikpivkin) [inline pkg/fanal/walker/vm.go:52]: Why are the skip directories from the option not used ?

**Review by simar7 — APPROVED** [review]

File: pkg/fanal/walker/fs_test.go (hunk 1 above)
Comment (knqyf263) [reply pkg/fanal/walker/fs_test.go:5]: Thanks. Yes, I usually use `slices`, but this PR was open on Sep 13, 2023 (and waiting for Aqua's review for a long time). Trivy was using Go 1.20, which doesn't have the `slices` package. I'll update it.

File: pkg/fanal/artifact/vm/vm.go [hunk 3]
```diff
@@ -15,7 +16,7 @@ import (
 	"github.com/aquasecurity/trivy/pkg/fanal/handler"
 	"github.com/aquasecurity/trivy/pkg/fanal/types"
 	"github.com/aquasecurity/trivy/pkg/fanal/walker"
-	"github.com/aquasecurity/trivy/pkg/semaphore"
+	semaphore "github.com/aquasecurity/trivy/pkg/semaphore"
```
Comment (DmitriyLewen) [inline pkg/fanal/artifact/vm/vm.go:19]: I'm just curious 😄
It looks like you renamed `semaphore` to `semaphore`.
Is there some reason for this?

**Review by DmitriyLewen — APPROVED** [review]
LGTM

File: pkg/fanal/walker/walk.go [hunk 4]
```diff
@@ -6,85 +6,47 @@ import (
...
-const defaultSizeThreshold = int64(200) << 20 // 200MB
+const defaultSizeThreshold = int64(100) << 20 // 200MB
 
-type WalkFunc func(filePath string, info os.FileInfo, opener analyzer.Opener) error
-
-type walker struct {
-	skipFiles []string
-	skipDirs  []string
+var defaultSkipDirs = []string{
+	"**/.git",
```
Comment (Esonhugh) [inline pkg/fanal/walker/walk.go:18]: I have a question. Why you just ignore the .git folder?

Comment (knqyf263) [reply pkg/fanal/walker/walk.go:18]: The git dir shouldn't have files Trivy wants to find, and it is time-consuming to walk the dir.

Comment (Esonhugh) [reply pkg/fanal/walker/walk.go:18]: > The git dir shouldn't have files Trivy wants to find, and it is time-consuming to walk the dir.

thank u for answering. But I want to say something. just like what github official does.

https://github.com/actions/checkout/blob/44c2b7a8a4ea60a981eaca3cf939b5f4305c123b/src/git-auth-helper.ts#L286

They allows save PAT tokens or request OAuth tokens in this actions. And backup this  So some careless people do like 

```dockerfile
FROM node:16
RUN mkdir /app
COPY . /app
RUN yarn install
EXPOSE 80
CMD ["node", "/app/app.js"]
```
And Python/PHP/Javascript programmers always do like this. : ) 

So at least whitelist the `.git/config` `.git/modules/[any]/config` and blacklist something like `.git/objects`

Comment (knqyf263) [reply pkg/fanal/walker/walk.go:18]:
```
FROM node:16
RUN mkdir /app
COPY . /app
RUN yarn install
EXPOSE 80
CMD ["node", "/app/app.js"]
```

What's wrong in this Dockerfile? PAT will be deleted in the post-job step.

Comment (Esonhugh) [reply pkg/fanal/walker/walk.go:18]: > ```
> FROM node:16
> RUN mkdir /app
> COPY . /app
> RUN yarn install
> EXPOSE 80
> CMD ["node", "/app/app.js"]
> ```
> 
> What's wrong in this Dockerfile? PAT will be deleted in the post-job step.

Yes post job. But docker image building is before it :) 
It's 

```mermaid
graph TD;
START ---> checkout ---> docker-image-building ---> docker-image-uploading ---> post-docker-login ----> post-checkout  ---> deleted-PAT ---> github.token-expired
```

Comment (knqyf263) [reply pkg/fanal/walker/walk.go:18]: Hmm. It's interesting. Did you ensure PAT can be exposed to a container image?

Comment (ActionOPs) [reply pkg/fanal/walker/walk.go:18]: > Hmm. It's interesting. Did you ensure PAT can be exposed to a container image?

pretty sure. Play with cross repos ref submodules  they must do like that. I can create an environment. : )

Comment (knqyf263) [reply pkg/fanal/walker/walk.go:18]: I've created an issue. Thanks!
https://github.com/aquasecurity/trivy/issues/6699

Comment (Esonhugh) [reply pkg/fanal/walker/walk.go:18]: > I've created an issue. Thanks! #6699

nice work :)

**Pull request merged**